crates/sui/src/client_commands.rs

```diff
@@ -840,6 +851,47 @@ impl SuiClientCommands {
                 };
                 SuiClientCommandResult::Addresses(output)
             }
+            SuiClientCommands::Balance {
+                address,
+                with_coins,
+            } => {
+                let address = get_identity_address(address, context)?;
+                let client = context.get_client().await?;
+                let mut objects: Vec<Coin> = Vec::new();
+                let mut cursor = None;
+                loop {
+                    let response = client
+                        .coin_read_api()
+                        .get_all_coins(address, cursor, None)
+                        .await?;
+                    objects.extend(response.data);
+
+                    if response.has_next_page {
+                        cursor = response.next_cursor;
+                    } else {
+                        break;
+                    }
+                }
+                let mut coins_by_type: BTreeMap<String, (Option<SuiCoinMetadata>, Vec<Coin>)> =
+                    BTreeMap::new();
+                for c in objects {
```

Comment:
it seems like we're fetching the `coin_metadata` for each coin object, so potentially we may fetch the same `coin_metadata` multiple times. It also forces you to store `SuiCoinMetadata` as an `Option` in the tuple. What about keeping two maps, `BTreeMap<String, Vec<Coin>>` and `BTreeMap<String, CoinMetadata>`? 

for each coin, check if there's an entry in the second BTreeMap, if not, then we can fetch. This way you don't have to deal with `Option<SuiCoinMetadata>` and you can avoid redundant fetches

Reply:
I don't think you end up requesting the metadata multiple times because you fetch it only if the entry doesn't exist in the `BTreeMap`.

Reply:
This shouldn't be redundant, but I will double check. Thanks!